core/services/vrf/delegate.go

```diff
@@ -122,24 +137,59 @@ var (
 	_ job.Service  = &listener{}
 )
 
+type request struct {
+	confirmedAtBlock uint64
+	req              *solidity_vrf_coordinator_interface.VRFCoordinatorRandomnessRequest
+	lb               log.Broadcast
+}
+
 type listener struct {
-	cfg            Config
-	l              logger.Logger
-	abi            abi.ABI
-	logBroadcaster log.Broadcaster
-	coordinator    *solidity_vrf_coordinator_interface.VRFCoordinator
-	pipelineRunner pipeline.Runner
-	pipelineORM    pipeline.ORM
-	vorm           keystore.VRFORM
-	job            job.Job
-	db             *gorm.DB
-	txm            bulletprooftxmanager.TxManager
-	vrfks          *keystore.VRF
-	gethks         GethKeyStore
-	mbLogs         *utils.Mailbox
-	chStop         chan struct{}
-	waitOnStop     chan struct{}
 	utils.StartStopOnce
+
+	cfg             Config
+	l               logger.Logger
+	abi             abi.ABI
+	logBroadcaster  log.Broadcaster
+	coordinator     *solidity_vrf_coordinator_interface.VRFCoordinator
+	pipelineRunner  pipeline.Runner
+	pipelineORM     pipeline.ORM
+	vorm            keystore.VRFORM
+	job             job.Job
+	db              *gorm.DB
+	headBroadcaster httypes.HeadBroadcasterRegistry
+	txm             bulletprooftxmanager.TxManager
+	vrfks           *keystore.VRF
+	gethks          GethKeyStore
+	reqLogs         *utils.Mailbox
+	chStop          chan struct{}
+	waitOnStop      chan struct{}
+	newHead         chan struct{}
+	latestHead      uint64 // Only one writer and one reader, no lock needed
+	// We can keep these pending logs in memory because we
+	// only mark them confirmed once we send a corresponding fulfillment transaction.
+	// So on node restart in the middle of processing, the lb will resend them.
+	reqsMu   sync.Mutex // Both goroutines write to reqs
+	reqs     []request
+	reqAdded func() // A simple debug helper
+
+	// Data structures for reorg attack protection
+	respCountMu        sync.Mutex
+	respCount          map[[32]byte]uint64
+	blockNumberToReqID *pairing.PairHeap
+}
+
+func (lsn *listener) Connect(head *models.Head) error {
+	lsn.latestHead = uint64(head.Number)
+	return nil
+}
+
+// Note that we have 2 seconds to do this processing
+func (lsn *listener) OnNewLongestChain(ctx context.Context, head models.Head) {
+	lsn.latestHead = uint64(head.Number)
```

Comment:
hm seems like `lsn.latestHead` is accessed in multiple goroutines - might need to be an atomic?